lib/mixins/properties-mixin.html

```diff
@@ -0,0 +1,212 @@
+      for (let p in props) {
+        let o = props[p];
+        if (typeof o == 'function') {
+          o = { type: o };
+        }
+        output[p] = o;
+      }
+      return output;
+    }
+
+    /**
+     * Returns the super class constructor for the given class, if it is an
+     * instance of the PropertiesMixin.
+     *
+     * @param {PropertiesMixinConstructor} constructor PropertiesMixin constructor
+     * @return {PropertiesMixinConstructor} Super class constructor
+     */
+    function superForClass(constructor) {
+      const proto = /** @type {PropertiesMixinConstructor} */ (constructor).prototype;
+      const superCtor = Object.getPrototypeOf(proto).constructor;
+      if (superCtor.prototype instanceof PropertiesMixin) {
+        return superCtor;
+      }
+    }
+
+    /**
+     * Returns a memoized version of the `properties` object for the
+     * given class. Properties not in object format are converted to at
+     * least {type}.
+     *
+     * @param {PropertiesMixinConstructor} constructor PropertiesMixin constructor
+     * @return {Object} Memoized properties object
+     */
+    function ownProperties(constructor) {
+      if (!constructor.hasOwnProperty(
+        JSCompiler_renameProperty('__ownProperties', constructor))) {
+        const props = constructor.properties;
+        constructor.__ownProperties = props ? normalizeProperties(props) : null;
+      }
+      return constructor.__ownProperties;
+    }
+
+    /**
+     * @polymer
+     * @mixinClass
+     * @extends {base}
+     * @implements {Polymer_PropertiesMixin}
+     * @unrestricted
+     */
+    class PropertiesMixin extends base {
+
+      /**
+       * Implements standard custom elements getter to observes the attributes
+       * listed in `properties`.
+       */
+      static get observedAttributes() {
+        const props = this._properties;
+        return props ? Object.keys(props).map(p => {
+          return this.attributeNameForProperty(p);
+        }) : [];
+      }
+
+      /**
+       * Finalizes an element definition, including ensuring any super classes
+       * are also finalized. This includes ensuring property
+       * accessors exist on the element prototype. This method calls
+       * `_finalizeClass` to finalize each constructor in the prototype chain.
+       */
+       static finalize() {
+        if (!this.hasOwnProperty(JSCompiler_renameProperty('__finalized', this))) {
+          const superCtor = superForClass(this);
+          if (superCtor) {
+            superCtor.finalize();
+          }
+          this.__finalized = true;
+          this._finalizeClass();
+        }
+      }
+
+      /**
+       * Finalize an element class. This includes ensuring property
+       * accessors exist on the element prototype. This method is called by
+       * `finalize` and finalizes the class constructor.
+       *
+       * @protected
+       */
+      static _finalizeClass() {
+        const props = ownProperties(this);
```

Comment:
Looked at this with Justin, the answer is because `observedAttributes` needs access to properties, and want to keep that as minimal as possible.

Reply:
But agree `ownProperties` function & memoizer could be removed and just have own properties added in `get _properties()` and use `Object.keys()` to get "own properties".

Reply:
We need to pass only "own properties" to `createProperties`. To do that with only `_properties`, we'd have to copy the top level properties into another object which is more code and slower. While we could change `createProperties` to only create properties for `ownPropertyNames` that seems like an odd choice.

Also, we have `ownProperties` to avoid hitting the `properties` getter more than once (to avoid running potentially $ user code). This getter can be called in one of two cases, (1) define is called and `_properties` is accessed via `observedAttributes` being inspected; (2) the class is "finalized" without being defined by virtue of being a superclass.

This is definitely biased for avoiding runtime work v. minimizing code size.

Reply:
I don't think it's _that_ odd to define the `_properties` object that this generates and `createProperties` accepts as have a distinction between own and inherited properties :) The behavior would be that the _normalized_ properties object has a prototype chain that mirrors the prototype chain of the constructor, just like the proto chain of the constructors mirrors the proto chain of their `.prototype` properties. Makes sense to me, but I admit it might not for everyone.

Arguing against myself out-loud: the difference between `Object.keys(this._properties)` and `for (let p in this._properties) {}` might be way too subtle. It's also admittedly a very minor benefit, because it's per-class...

...but it is there: slightly less code, properties are only looped through once to build `__properties`, less duplication of key/values between `__ownProperties` and `__properties`, one less static property and shape change.

Anyway, super, super small nit. I'm hyper-zoomed in here, apologies...